### PR TITLE
implement elliptic-curve point addition for general rings

### PR DESCRIPTION
**Pull request by yyyyx4**
Here we revive the effort from #33228 to implement point-addition formulas on elliptic curves over (more) general rings.

This resolves #33228.

### CODE REVIEW

File: src/sage/schemes/elliptic_curves/addition_formulas_ring.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    a1, a2, a3, a4, a6 = E.a_invariants()
+    b2, b4, b6, b8 = E.b_invariants()
+
+    assert P in E
+    assert Q in E
+    X1, Y1, Z1 = P
+    X2, Y2, Z2 = Q
+
+    #TODO: I've made a half-hearted attempt at simplifying the formulas
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/addition_formulas_ring.py]:
```suggestion
    # TODO: I've made a half-hearted attempt at simplifying the formulas
```

File: src/sage/schemes/elliptic_curves/ell_generic.py [hunk 2]
```diff
@@ -175,15 +173,46 @@ def __init__(self, K, ainvs, category=None):
...
     _point = ell_point.EllipticCurvePoint
 
+    def assume_base_ring_is_field(self, flag=True):
+        r"""
+        Set a flag to pretend that this elliptic curve is defined over a
+        field while doing arithmetic, which is useful in some algorithms.
+
+        The flag affects all points created while the flag is set. Note
+        that elliptic curves are unique parents, hence setting this flag
+        may break seemingly unrelated parts of Sage.
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/ell_generic.py]:
```suggestion

        .. WARNING::
        
            The flag affects all points created while the flag is set. Note
            that elliptic curves are unique parents, hence setting this flag
            may break seemingly unrelated parts of Sage.
```

Comment (vincentmacri) [reply src/sage/schemes/elliptic_curves/ell_generic.py]: Breaking other parts of Sage seems bad, so I suggest adding a WARNING block for emphasis

File: src/sage/schemes/elliptic_curves/ell_generic.py [hunk 3]
```diff
@@ -175,15 +173,46 @@ def __init__(self, K, ainvs, category=None):
...
+            sage: 4*P
+            (23 : 26 : 1)
+            sage: 9*P
+            Traceback (most recent call last):
+            ...
+            ZeroDivisionError: Inverse of 5 does not exist (characteristic = 35 = 5*7)
+
+        .. NOTE::
+
+            This method is a **hack** provided for educational purposes.
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/ell_generic.py]:
```suggestion
```
I think it makes sense to have the explanation of the function before the examples.

File: src/sage/schemes/elliptic_curves/ell_generic.py (hunk 2 above)
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/ell_generic.py]:
```suggestion
        may break seemingly unrelated parts of Sage.

        .. NOTE::
            This method is a **hack** provided for educational purposes.

```

File: src/sage/schemes/elliptic_curves/ell_generic.py [hunk 4]
```diff
@@ -582,7 +611,7 @@ def __call__(self, *args, **kwds):
             # infinity.
             characteristic = self.base_ring().characteristic()
             if characteristic != 0 and isinstance(args[0][0], Rational) and isinstance(args[0][1], Rational):
-                if mod(args[0][0].denominator(),characteristic) == 0 or mod(args[0][1].denominator(),characteristic) == 0:
+                if characteristic.divides(args[0][0].denominator()) or characteristic.divides(args[0][1].denominator()):
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/ell_generic.py:617]: Out of curiosity, what's the reason for this change? It is more efficient?

File: src/sage/schemes/elliptic_curves/ell_point.py [hunk 5]
```diff
@@ -73,15 +75,24 @@
     sage: P*(n+1)-P*n == P
     True
 
-Arithmetic over `\ZZ/N\ZZ` with composite `N` is supported.  When an
-operation tries to invert a non-invertible element, a
-:exc:`ZeroDivisionError` is raised and a factorization of the modulus appears
-in the error message::
+Arithmetic over `\ZZ/N\ZZ` with composite `N` is supported::
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/ell_point.py]:
```suggestion
Arithmetic over `\Zmod{N}` with composite `N` is supported::
```
Preferred for readability: https://doc.sagemath.org/html/en/developer/coding_basics.html#latex-typesetting

**Review by vincentmacri — CHANGES_REQUESTED** [review]
First review pass. I was only checking for style/formatting. I'll find some time to look at the math later, this looks like a cool contribution.

A few of the test failures seem relevant. Can you investigate and fix them?

File: src/sage/schemes/elliptic_curves/ell_generic.py (hunk 4 above)
Comment (yyyyx4) [reply src/sage/schemes/elliptic_curves/ell_generic.py:617]: Yes, it performs the divisibility test directly on integers, hence avoids the type conversions for the `mod()`, and this leads to a $\approx100\times$ speedup. The main reason was that it's the "morally" correct way of writing the test, though.

File: src/doc/en/reference/references/index.rst [hunk 6]
```diff
@@ -970,6 +973,10 @@ REFERENCES:
             Anal. Appl. 15 (1994) 804-823.
             :doi:`10.1137/S0895479892230031`
 
+.. [BL1995] W. Bosma, H.W. Lenstra: Complete Systems of Addition Laws for
```
Comment (vincentmacri) [inline src/doc/en/reference/references/index.rst]:
```suggestion
.. [BL1995] W. Bosma, H.W. Lenstra: Complete Systems of Two Addition Laws for
```
I think this is the name of the paper? Unless I'm looking at the wrong thing.

File: src/sage/schemes/elliptic_curves/ell_generic.py [hunk 7]
```diff
@@ -175,15 +173,49 @@ def __init__(self, K, ainvs, category=None):
...
+        r"""
+        Set a flag to pretend that this elliptic curve is defined over a
+        field while doing arithmetic, which is useful in some algorithms.
+
+
+        .. WARNING::
+
+            The flag affects all points created while the flag is set. Note
+            that elliptic curves are unique parents, hence setting this flag
+            may break seemingly unrelated parts of Sage.
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/ell_generic.py:188]: > setting this flag may break seemingly unrelated parts of Sage

This makes me uncomfortable. Do we have anything similar in other parts of Sage? How necessary/useful is this function?

File: src/sage/schemes/elliptic_curves/addition_formulas_ring.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    a1sq, a2sq, a3sq, a4sq = (a**2 for a in (a1, a2, a3, a4))
+
+    X31 = XYdif*YZsum+XZdif*Y1*Y2+a1*X1*X2*YZdif+a1*XYdif*XZsum-a2*X1*X2*XZdif+a3*XYdif*Z1*Z2+a3*XZdif*YZsum-a4*XZsum*XZdif-3*a6*XZdif*Z1*Z2
+
+    Y31 = -3*X1*X2*XYdif-Y1*Y2*YZdif-2*a1*XZdif*Y1*Y2+(a1sq+3*a2)*X1*X2*YZdif-(a1sq+a2)*XYsum*XZdif+(a1*a2-3*a3)*X1*X2*XZdif-(2*a1*a3+a4)*XYdif*Z1*Z2+a4*XZsum*YZdif+(a1*a4-a2*a3)*XZsum*XZdif+(a3sq+3*a6)*YZdif*Z1*Z2+(3*a1*a6-a3*a4)*XZdif*Z1*Z2
+
+    Z31 = 3*X1*X2*XZdif-YZsum*YZdif+a1*XYdif*Z1*Z2-a1*XZdif*YZsum+a2*XZsum*XZdif-a3*YZdif*Z1*Z2+a4*XZdif*Z1*Z2
+
+    yield (X31, Y31, Z31)
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/addition_formulas_ring.py:81]:
```suggestion
```

Using `yield` makes this function a generator which leads to some unintuitive behaviour.

For instance,

```python
p_plus_q = add(E, P, Q)
PQ1, PQ2 = p_plus_q
PQ3, PQ4 = p_plus_q  # This will give an error since the generator is exhausted
```

File: src/sage/schemes/elliptic_curves/addition_formulas_ring.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+    Z32 = 3*X1*X2*XYsum+Y1*Y2*YZsum+3*a1*X1**2*X2**2+a1*(2*X1*Y2+Y1*X2)*Y1*Z2+a1sq*X1*Z2*(2*X2*Y1+X1*Y2)+a2*X1*X2*YZsum+a2*XYsum*XZsum+a1**3*X1**2*X2*Z2+a1*a2*X1*X2*(2*X1*Z2+X2*Z1)+3*a3*X1*X2**2*Z1+a3*Y1*Z2*(YZsum+Y2*Z1)+2*a1*a3*X1*Z2*YZsum+2*a1*a3*X2*Y1*Z1*Z2+a4*XYsum*Z1*Z2+a4*XZsum*YZsum+(a1sq*a3+a1*a4)*X1*Z2*(XZsum+X2*Z1)+a2*a3*X2*Z1*(2*X1*Z2+X2*Z1)+a3sq*Y1*Z1*Z2**2+(a3sq+3*a6)*YZsum*Z1*Z2+a1*a3sq*(2*X1*Z2+X2*Z1)*Z1*Z2+3*a1*a6*X1*Z1*Z2**2+a3*a4*(XZsum+X2*Z1)*Z1*Z2+(a3**3+3*a3*a6)*Z1**2*Z2**2
+
+    yield (X32, Y32, Z32)
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/addition_formulas_ring.py:89]:
```suggestion
    return (X31, Y31, Z31), (X32, Y32, Z32)
```

File: src/sage/schemes/elliptic_curves/addition_formulas_ring.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
+
+def add(E, P, Q):
+    r"""
+    Addition formulas for elliptic curves over general rings
+    with trivial Picard group.
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/addition_formulas_ring.py:5]: Add input and output blocks: https://doc.sagemath.org/html/en/developer/coding_basics.html#documentation-strings.

Can you include a brief explanation of why this function returns two values? I would expect an "add" function would only return one value, so this request for additional explanation is also partly to help me review the code.

File: src/sage/schemes/elliptic_curves/addition_formulas_ring.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        sage: PQ2
+        (582, 2347, 1028)
+        sage: E(PQ1) == E(PQ2)
+        True
+    """
+    a1, a2, a3, a4, a6 = E.a_invariants()
+    b2, b4, b6, b8 = E.b_invariants()
+
+    assert P in E
+    assert Q in E
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/addition_formulas_ring.py]:
```suggestion
    if P not in E:
        raise ValueError('P must be in E')
    if Q not in E:
        raise ValueError('Q must be in E')
```
Don't use asserts to validate input.

File: src/sage/schemes/projective/projective_point.py [hunk 12]
```diff
@@ -533,7 +533,11 @@ def scale_by(self, t):
 
     def normalize_coordinates(self):
         """
-        Removes the gcd from the coordinates of this point (including `-1`).
+        Removes the gcd from the coordinates of this point (including `-1`)
```
Comment (vincentmacri) [inline src/sage/schemes/projective/projective_point.py:536]: I like this change.

File: src/doc/en/reference/references/index.rst (hunk 6 above)
Comment (yyyyx4) [reply src/doc/en/reference/references/index.rst]: Indeed, good catch, thanks.

File: src/sage/schemes/elliptic_curves/addition_formulas_ring.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        sage: E = EllipticCurve(S, [1,2,3,4,5])
+        sage: P = E(817, 13, 19)
+        sage: Q = E(425, 123, 17)
+        sage: PQ1, PQ2 = add(E, P, Q)
+        sage: PQ1
+        (1188, 1674, 540)
+        sage: PQ2
+        (582, 2347, 1028)
+        sage: E(PQ1) == E(PQ2)
+        True
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/addition_formulas_ring.py:36]: Given the length of the formulas here, can we include a test that when you use this additional algorithm on a curve defined over a field, we get the same thing as the regular addition formulas for an elliptic curve over a field?

**Review by vincentmacri — CHANGES_REQUESTED** [review]
Some additional comments, requests for more detailed docstrings, etc.

Unless I missed it, it seems all your tests for this are over Zmod(N)? Since this is supposed to work for general rings, can you include a test with a more complicated ring?

File: src/sage/schemes/elliptic_curves/ell_point.py [hunk 14]
```diff
@@ -179,6 +220,259 @@ def curve(self):
...
+
+        from sage.schemes.elliptic_curves.addition_formulas_ring import add
+        from sage.modules.free_module_element import vector
+
+        pts = []
+        for pt in filter(any, add(E, self, other)):
+            if R.one() in R.ideal(pt):
+                return E.point(pt)
+            pts.append(pt)
+        assert len(pts) == 2, 'bug in elliptic-curve point addition'
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/ell_point.py:356]: Raise an error instead of using an assert

File: src/sage/schemes/elliptic_curves/ell_point.py [hunk 15]
```diff
@@ -179,6 +220,259 @@ def curve(self):
...
+        # find a good choice. Is there a general method that doesn't
+        # involve guessing?
+
+        pts = [vector(R, pt) for pt in pts]
+        for _ in range(1000):
+            result = tuple(sum(R.random_element() * pt for pt in pts))
+            if R.one() in R.ideal(result):
+                return E.point(result, check=False)
+
+        assert False, 'bug: failed to compute elliptic-curve point addition'
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/ell_point.py:371]: Raise an error instead of using an assert

File: src/sage/schemes/elliptic_curves/ell_point.py [hunk 16]
```diff
@@ -727,72 +1027,52 @@ def _add_(self, right):
...
-                    N1 = N.gcd(Integer(x1-x2))
-                    N2 = N//N1
-                    raise ZeroDivisionError("Inverse of %s does not exist (characteristic = %s = %s*%s)" % (x1-x2, N, N1, N2))
-                else:
-                    raise ZeroDivisionError("Inverse of %s does not exist" % (x1-x2))
+                d = next(d for d in (x1 - x2, 2*y1 + a1*x1 + a3) if d and not d.is_unit())
+                m, = d.parent().defining_ideal().gens()
+                f1 = d.lift().gcd(m)
+                f2 = m // f1
+                assert m == f1 * f2
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/ell_point.py:1056]: Raise an error instead of using an assert

**Review by vincentmacri — CHANGES_REQUESTED** [review]
Additional comments for code cleanup

File: src/sage/schemes/elliptic_curves/ell_generic.py (hunk 7 above)
Comment (yyyyx4) [reply src/sage/schemes/elliptic_curves/ell_generic.py:188]: I was going to remove this hack altogether, but over at #33228 some people insisted that it remain...

File: src/sage/schemes/elliptic_curves/addition_formulas_ring.py (hunk 8 above)
Comment (yyyyx4) [reply src/sage/schemes/elliptic_curves/addition_formulas_ring.py:81]: This function is not intended for users. They should just write `P + Q` instead.

File: src/sage/schemes/elliptic_curves/ell_point.py (hunk 14 above)
Comment (yyyyx4) [reply src/sage/schemes/elliptic_curves/ell_point.py:356]: No, this one really is an assertion: If it fails, there is a bug. That's exactly what assertions are for.

File: src/sage/schemes/elliptic_curves/ell_point.py (hunk 15 above)
Comment (yyyyx4) [reply src/sage/schemes/elliptic_curves/ell_point.py:371]: Same here.

File: src/sage/schemes/elliptic_curves/ell_point.py (hunk 16 above)
Comment (yyyyx4) [reply src/sage/schemes/elliptic_curves/ell_point.py:1056]: Same here.

File: src/sage/schemes/elliptic_curves/addition_formulas_ring.py (hunk 9 above)
Comment (yyyyx4) [reply src/sage/schemes/elliptic_curves/addition_formulas_ring.py:89]: This is less efficient: If the first formula works, there is no point in evaluating the second one.

File: src/sage/schemes/elliptic_curves/addition_formulas_ring.py (hunk 10 above)
Comment (yyyyx4) [reply src/sage/schemes/elliptic_curves/addition_formulas_ring.py:5]: I'll add an explanation to the code momentarily. The point is that there are two separate addition formulas, which are equivalent whenever both work, but it is only guaranteed that some linear combination of them works for any given set of inputs. Hence why it yields the results for both possible formulas, and the caller will opportunistically take the first result that works, or combine both to obtain a valid projective point.

File: src/sage/schemes/elliptic_curves/addition_formulas_ring.py (hunk 8 above)
Comment (vincentmacri) [reply src/sage/schemes/elliptic_curves/addition_formulas_ring.py:81]: Okay. If this is meant to be internal then perhaps rename to `_add`? Names starting with an underscore are considered internal: https://doc.sagemath.org/html/en/developer/coding_in_python.html#deprecation

File: src/sage/schemes/elliptic_curves/addition_formulas_ring.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
+
+def add(E, P, Q):
```
Comment (vincentmacri) [inline src/sage/schemes/elliptic_curves/addition_formulas_ring.py]:
```suggestion
def _add(E, P, Q):
```
Internal functions should start with an underscore, per the recently-clarified deprecation policy: https://groups.google.com/g/sage-devel/c/7T0Ys01Kcsg/m/23Rcg8uzAwAJ

File: src/sage/schemes/elliptic_curves/ell_generic.py (hunk 7 above)
Comment (vincentmacri) [reply src/sage/schemes/elliptic_curves/ell_generic.py:188]: Okay. We can leave it for now.

**Comment by yyyyx4** [comment]
Alright, I believe all your comments have been dealt with now. Thanks for the thorough review!

**Review by vincentmacri — APPROVED** [review]
Looks good to me. Thanks for the contribution!

**Comment by user202729** [comment]
This makes meson build fails. (I guess the fact that meson build fails all the time makes nobody want to look at it)

```diff
diff --git a/src/sage/schemes/elliptic_curves/meson.build b/src/sage/schemes/elliptic_curves/meson.build
index 3448c5d1c0a..5f12079dc03 100644
--- a/src/sage/schemes/elliptic_curves/meson.build
+++ b/src/sage/schemes/elliptic_curves/meson.build
@@ -1,6 +1,7 @@
 py.install_sources(
   'BSD.py',
   'Qcurves.py',
+  'addition_formulas_ring.py',
   'all.py',
   'cardinality.py',
   'cm.py',
```

Not sure if someone already made a pull request though.
